File: app/src/main/java/com/parabola/newtone/mvp/presenter/TabPlaylistPresenter.java
package com.parabola.newtone.mvp.presenter;

import com.parabola.domain.executor.SchedulerProvider;
import com.parabola.domain.interactor.player.PlayerInteractor;
import com.parabola.domain.repository.PlaylistRepository;
import com.parabola.domain.repository.TrackRepository;
import com.parabola.domain.settings.ViewSettingsInteractor;
import com.parabola.newtone.di.app.AppComponent;
import com.parabola.newtone.mvp.view.TabPlaylistView;
import com.parabola.newtone.ui.router.MainRouter;

import javax.inject.Inject;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.internal.functions.Functions;
import io.reactivex.internal.observers.ConsumerSingleObserver;
import moxy.InjectViewState;
import moxy.MvpPresenter;

@InjectViewState
public final class TabPlaylistPresenter extends MvpPresenter<TabPlaylistView> {

    @Inject MainRouter router;

    @Inject PlayerInteractor playerInteractor;
    @Inject PlaylistRepository playlistRepo;
    @Inject TrackRepository trackRepo;
    @Inject SchedulerProvider schedulers;
    @Inject ViewSettingsInteractor viewSettingsInteractor;

    private final CompositeDisposable disposables = new CompositeDisposable();

    public TabPlaylistPresenter(AppComponent appComponent) {
        appComponent.inject(this);
    }

    @Override
    protected void onFirstViewAttach() {
        disposables.addAll(
                observePlaylistsUpdates(),
                observeIsItemDividerShowed(),
                observeTrackDeleting());
    }

    @Override
    public void onDestroy() {
        disposables.dispose();
    }

    private Disposable observePlaylistsUpdates() {
        return playlistRepo.observePlaylistsUpdates()
                .flatMapSingle(o -> playlistRepo.getAll())
                .subscribeOn(schedulers.io())
                .observeOn(schedulers.ui())
                .subscribe(getViewState()::refreshPlaylists);
    }

    private Disposable observeIsItemDividerShowed() {
        return viewSettingsInteractor.observeIsItemDividerShowed()
                .subscribe(getViewState()::setItemDividerShowing);
    }

    private Disposable observeTrackDeleting() {
        return trackRepo.observeTrackDeleting()
                .flatMapSingle(deletedTrackId -> playlistRepo.getAll())
                .observeOn(schedulers.ui())
                .subscribe(getViewState()::refreshPlaylists);
    }

    public void onClickPlaylistItem(int selectedPlaylistId) {
        router.openPlaylist(selectedPlaylistId);
    }

    public void onClickMenuRename(int playlistId) {
        router.openRenamePlaylistDialog(playlistId);
    }

    public void onClickMenuShuffle(int playlistId) {
        trackRepo.getByPlaylist(playlistId)
                .subscribe(new ConsumerSingleObserver<>(
                        playerInteractor::startInShuffleMode,
                        Functions.ERROR_CONSUMER));
    }

    public void onClickMenuDelete(int deletedPlaylistId) {
        playlistRepo.remove(deletedPlaylistId)
                .subscribe();
    }

    public void onClickRecentlyAdded() {
        router.openRecentlyAdded();
    }

    public void onClickFavourites() {
        router.openFavourites();
    }

    public void onClickQueue() {
        router.openQueue();
    }

    public void onClickFolders() {
        router.openFoldersList();
    }

}

File: app/src/main/java/com/parabola/newtone/presentation/main/playlists/TabPlaylistPresenter.kt
package com.parabola.newtone.presentation.main.playlists

import com.parabola.domain.executor.SchedulerProvider
import com.parabola.domain.interactor.player.PlayerInteractor
import com.parabola.domain.repository.PlaylistRepository
import com.parabola.domain.repository.TrackRepository
import com.parabola.domain.settings.ViewSettingsInteractor
import com.parabola.newtone.di.app.AppComponent
import com.parabola.newtone.ui.router.MainRouter
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.internal.functions.Functions
import io.reactivex.internal.observers.ConsumerSingleObserver
import moxy.InjectViewState
import moxy.MvpPresenter
import javax.inject.Inject

@InjectViewState
class TabPlaylistPresenter(appComponent: AppComponent) : MvpPresenter<TabPlaylistView>() {

    @Inject
    lateinit var router: MainRouter

    @Inject
    lateinit var playerInteractor: PlayerInteractor

    @Inject
    lateinit var playlistRepo: PlaylistRepository

    @Inject
    lateinit var trackRepo: TrackRepository

    @Inject
    lateinit var schedulers: SchedulerProvider

    @Inject
    lateinit var viewSettingsInteractor: ViewSettingsInteractor

    private val disposables = CompositeDisposable()


    init {
        appComponent.inject(this)
    }


    override fun onFirstViewAttach() {
        disposables.addAll(
            observePlaylistsUpdates(),
            observeIsItemDividerShowed(),
            observeTrackDeleting()
        )
    }

    override fun onDestroy() {
        disposables.dispose()
    }


    private fun observePlaylistsUpdates(): Disposable {
        return playlistRepo.observePlaylistsUpdates()
            .flatMapSingle { playlistRepo.all }
            .subscribeOn(schedulers.io())
            .observeOn(schedulers.ui())
            .subscribe(viewState::refreshPlaylists)
    }

    private fun observeIsItemDividerShowed(): Disposable {
        return viewSettingsInteractor.observeIsItemDividerShowed()
            .subscribe(viewState::setItemDividerShowing)
    }

    private fun observeTrackDeleting(): Disposable {
        return trackRepo.observeTrackDeleting()
            .flatMapSingle { playlistRepo.all }
            .observeOn(schedulers.ui())
            .subscribe(viewState::refreshPlaylists)
    }


    fun onClickPlaylistItem(selectedPlaylistId: Int) {
        router.openPlaylist(selectedPlaylistId)
    }

    fun onClickMenuRename(playlistId: Int) {
        router.openRenamePlaylistDialog(playlistId)
    }

    fun onClickMenuShuffle(playlistId: Int) {
        trackRepo.getByPlaylist(playlistId)
            .subscribe(
                ConsumerSingleObserver(
                    playerInteractor::startInShuffleMode,
                    Functions.ERROR_CONSUMER
                )
            )
    }

    fun onClickMenuDelete(deletedPlaylistId: Int) {
        playlistRepo.remove(deletedPlaylistId)
            .subscribe()
    }

    fun onClickRecentlyAdded() {
        router.openRecentlyAdded()
    }

    fun onClickFavourites() {
        router.openFavourites()
    }

    fun onClickQueue() {
        router.openQueue()
    }

    fun onClickFolders() {
        router.openFoldersList()
    }

}
